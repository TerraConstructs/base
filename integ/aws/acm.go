package aws

import (
	"fmt"
	"time"

	"github.com/aws/aws-sdk-go/service/acm"
	"github.com/stretchr/testify/require"

	"github.com/gruntwork-io/terratest/modules/aws"
	"github.com/gruntwork-io/terratest/modules/logger"
	"github.com/gruntwork-io/terratest/modules/retry"
	"github.com/gruntwork-io/terratest/modules/testing"
)

// ref: https://github.com/gruntwork-io/terratest/blob/v0.47.1/modules/aws/acm.go

// Get Certificate Status
func GetAcmCertificateStatus(t testing.TestingT, awsRegion string, certArn string) string {
	status, err := GetAcmCertificateStatusE(t, awsRegion, certArn)
	if err != nil {
		t.Fatal(err)
	}
	return status
}

// GetAcmCertificateStatusE gets the ACM certificate status for the given certificate ARN in the given region.
func GetAcmCertificateStatusE(t testing.TestingT, awsRegion string, certArn string) (string, error) {
	acmClient, err := aws.NewAcmClientE(t, awsRegion)
	if err != nil {
		return "", err
	}

	result, err := acmClient.DescribeCertificate(&acm.DescribeCertificateInput{
		CertificateArn: &certArn,
	})
	if err != nil {
		return "", err
	}

	return *result.Certificate.Status, nil
}

// WaitForCertificateIssued waits for the certificate to be issued
func WaitForCertificateIssued(
	t testing.TestingT,
	certArn string,
	region string,
	maxRetries int,
	sleepBetweenRetries time.Duration,
) {
	err := WaitForCertificateIssuedE(t, certArn, region, maxRetries, sleepBetweenRetries)
	require.NoError(t, err)
}

// WaitForCertificateIssuedE waits for the ACM Certificate to be issued
func WaitForCertificateIssuedE(
	t testing.TestingT,
	certArn string,
	region string,
	maxRetries int,
	sleepBetweenRetries time.Duration,
) error {
	msg, err := retry.DoWithRetryE(
		t,
		fmt.Sprintf("Waiting for Certificate %s to be ISSUED.", certArn),
		maxRetries,
		sleepBetweenRetries,
		func() (string, error) {
			certStatus, err := GetAcmCertificateStatusE(t, region, certArn)
			if err != nil {
				return "", err
			}
			if certStatus != acm.CertificateStatusIssued {
				return "", NewCertificateNotIssuedError(certArn, certStatus)
			}
			return fmt.Sprintf("Certificate %s is now at desired status %s", certArn, acm.CertificateStatusIssued), nil
		},
	)
	logger.Log(t, msg)
	return err
}
