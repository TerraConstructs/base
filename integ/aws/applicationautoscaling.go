package aws

import (
	"context"
	"fmt"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/applicationautoscaling"
	"github.com/aws/aws-sdk-go-v2/service/applicationautoscaling/types"
	"github.com/stretchr/testify/require"

	terratestaws "github.com/gruntwork-io/terratest/modules/aws"
	"github.com/gruntwork-io/terratest/modules/testing"
)

// GetTableTrackingPolicy gets the target tracking policy for a DynamoDB table or errors if not found
func GetTableTrackingPolicy(t testing.TestingT, awsRegion string, resourceId string) *types.ScalingPolicy {
	policy, err := GetTableTrackingPolicyE(t, awsRegion, resourceId)
	require.NoError(t, err)
	return policy
}

// GetTableTrackingPolicy gets the target tracking policy for a DynamoDB table or returns an error if not found
func GetTableTrackingPolicyE(t testing.TestingT, awsRegion string, resourceId string) (*types.ScalingPolicy, error) {
	policies := GetScalingPolicies(t, awsRegion, "dynamodb")

	var targetTrackingPolicy *types.ScalingPolicy
	for _, policy := range policies {
		if *policy.ResourceId == resourceId &&
			policy.ScalableDimension == types.ScalableDimensionDynamoDBTableReadCapacityUnits &&
			policy.PolicyType == types.PolicyTypeTargetTrackingScaling {
			targetTrackingPolicy = &policy
			break
		}
	}
	if targetTrackingPolicy == nil {
		return nil, fmt.Errorf("no target tracking policy found for resource ID: %s", resourceId)
	}
	return targetTrackingPolicy, nil
}

// GetScalableTargets gets the Application Auto Scaling scalable targets for the given service namespace
func GetScalableTargets(t testing.TestingT, region string, serviceNamespace string) []types.ScalableTarget {
	targets, err := GetScalableTargetsE(t, region, serviceNamespace)
	require.NoError(t, err)
	return targets
}

// GetScalableTargetsE gets the Application Auto Scaling scalable targets for the given service namespace
func GetScalableTargetsE(t testing.TestingT, region string, serviceNamespace string) ([]types.ScalableTarget, error) {
	client, err := NewApplicationAutoScalingClientE(t, region)
	if err != nil {
		return nil, err
	}

	result, err := client.DescribeScalableTargets(context.Background(), &applicationautoscaling.DescribeScalableTargetsInput{
		ServiceNamespace: types.ServiceNamespace(serviceNamespace),
	})

	if err != nil {
		return nil, err
	}

	return result.ScalableTargets, nil
}

// GetScalingPolicies gets the Application Auto Scaling scaling policies for the given service namespace
func GetScalingPolicies(t testing.TestingT, region string, serviceNamespace string) []types.ScalingPolicy {
	policies, err := GetScalingPoliciesE(t, region, serviceNamespace)
	require.NoError(t, err)
	return policies
}

// GetScalingPoliciesE gets the Application Auto Scaling scaling policies for the given service namespace
func GetScalingPoliciesE(t testing.TestingT, region string, serviceNamespace string) ([]types.ScalingPolicy, error) {
	client, err := NewApplicationAutoScalingClientE(t, region)
	if err != nil {
		return nil, err
	}

	result, err := client.DescribeScalingPolicies(context.Background(), &applicationautoscaling.DescribeScalingPoliciesInput{
		ServiceNamespace: types.ServiceNamespace(serviceNamespace),
	})

	if err != nil {
		return nil, err
	}

	return result.ScalingPolicies, nil
}

// GetScheduledActions gets the Application Auto Scaling scheduled actions for the given service namespace
func GetScheduledActions(t testing.TestingT, region string, serviceNamespace string) []types.ScheduledAction {
	actions, err := GetScheduledActionsE(t, region, serviceNamespace)
	require.NoError(t, err)
	return actions
}

// GetScheduledActionsE gets the Application Auto Scaling scheduled actions for the given service namespace
func GetScheduledActionsE(t testing.TestingT, region string, serviceNamespace string) ([]types.ScheduledAction, error) {
	client, err := NewApplicationAutoScalingClientE(t, region)
	if err != nil {
		return nil, err
	}

	result, err := client.DescribeScheduledActions(context.Background(), &applicationautoscaling.DescribeScheduledActionsInput{
		ServiceNamespace: types.ServiceNamespace(serviceNamespace),
	})

	if err != nil {
		return nil, err
	}

	return result.ScheduledActions, nil
}

// GetScalableTargetsByResourceId gets scalable targets filtered by resource ID
func GetScalableTargetsByResourceId(t testing.TestingT, region string, serviceNamespace string, resourceId string) []types.ScalableTarget {
	targets, err := GetScalableTargetsByResourceIdE(t, region, serviceNamespace, resourceId)
	require.NoError(t, err)
	return targets
}

// GetScalableTargetsByResourceIdE gets scalable targets filtered by resource ID
func GetScalableTargetsByResourceIdE(t testing.TestingT, region string, serviceNamespace string, resourceId string) ([]types.ScalableTarget, error) {
	client, err := NewApplicationAutoScalingClientE(t, region)
	if err != nil {
		return nil, err
	}

	result, err := client.DescribeScalableTargets(context.Background(), &applicationautoscaling.DescribeScalableTargetsInput{
		ServiceNamespace: types.ServiceNamespace(serviceNamespace),
		ResourceIds:      []string{resourceId},
	})

	if err != nil {
		return nil, err
	}

	return result.ScalableTargets, nil
}

// GetScheduledActionsByResourceId gets scheduled actions filtered by resource ID
func GetScheduledActionsByResourceId(t testing.TestingT, region string, serviceNamespace string, resourceId string) []types.ScheduledAction {
	actions, err := GetScheduledActionsByResourceIdE(t, region, serviceNamespace, resourceId)
	require.NoError(t, err)
	return actions
}

// GetScheduledActionsByResourceIdE gets scheduled actions filtered by resource ID
func GetScheduledActionsByResourceIdE(t testing.TestingT, region string, serviceNamespace string, resourceId string) ([]types.ScheduledAction, error) {
	client, err := NewApplicationAutoScalingClientE(t, region)
	if err != nil {
		return nil, err
	}

	result, err := client.DescribeScheduledActions(context.Background(), &applicationautoscaling.DescribeScheduledActionsInput{
		ServiceNamespace: types.ServiceNamespace(serviceNamespace),
		ResourceId:       aws.String(resourceId),
	})

	if err != nil {
		return nil, err
	}

	return result.ScheduledActions, nil
}

// NewApplicationAutoScalingClient creates a new Application Auto Scaling client
func NewApplicationAutoScalingClient(t testing.TestingT, region string) *applicationautoscaling.Client {
	client, err := NewApplicationAutoScalingClientE(t, region)
	require.NoError(t, err)
	return client
}

// NewApplicationAutoScalingClientE creates a new Application Auto Scaling client
func NewApplicationAutoScalingClientE(t testing.TestingT, region string) (*applicationautoscaling.Client, error) {
	sess, err := terratestaws.NewAuthenticatedSession(region)
	if err != nil {
		return nil, err
	}
	return applicationautoscaling.NewFromConfig(*sess), nil
}
