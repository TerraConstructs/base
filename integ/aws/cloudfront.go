package aws

import (
	"context"
	"encoding/json"
	"fmt"
	"strconv"

	"github.com/aws/aws-sdk-go-v2/config"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/cloudfront"
	"github.com/aws/aws-sdk-go-v2/service/cloudfront/types"
	"github.com/gruntwork-io/terratest/modules/logger"
	"github.com/gruntwork-io/terratest/modules/testing"
	"github.com/hashicorp/go-multierror"
	"github.com/stretchr/testify/require"
)

type CloudFrontTestFunctionResult struct {
	Utilization  int // The amount of time that the function took to run as a percentage of the maximum allowed time.
	ErrorMessage *string
	// The event object returned by the function. For more information about the
	// structure of the event object, see [Event object structure]in the Amazon CloudFront Developer Guide.
	//
	// [Event object structure]: https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/functions-event-structure.html
	Output map[string]interface{}
	// Contains the log lines that the function wrote (if any) when running the test.
	ExecutionLogs []string
}

const invalidFunctionErrorPrefix = "The CloudFront function associated with the CloudFront distribution is invalid or could not run. Error: "

// responseValidator is a function that validates the response of a CloudFront function test.
type responseValidator func(*CloudFrontTestFunctionResult) error

// Tests a CloudFront function.
//
// To test a function, you provide an event object that represents an HTTP request
// or response that your CloudFront distribution could receive in production.
// CloudFront runs the function, passing it the event object that you provided, and
// returns the function's result (the modified event object) in the response. The
// response also contains function logs and error messages, if any exist. For more
// information about testing functions, see [Testing functions]in the Amazon CloudFront Developer
// Guide.
//
// To test a function, you provide the function's name and stage (DEVELOPMENT or LIVE)
// along with the event object. To get the function's name and version, you can use
// ListFunctions and DescribeFunction.
//
// [Testing functions]: https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/managing-functions.html#test-function
func TestCloudFrontFunction(t testing.TestingT, name string, stage string, event CloudFrontFunctionEvent, expectedError string, expectedOutput *map[string]interface{}) error {
	validateResponse := responseValidator(func(r *CloudFrontTestFunctionResult) error {
		var combinedErr error
		gotError := aws.ToString(r.ErrorMessage)
		if gotError != expectedError {
			combinedErr = multierror.Append(combinedErr, fmt.Errorf("got Error :%s\nWant Error:%s", gotError, expectedError))
		}

		if expectedOutput != nil {
			expectedOutputBytes, _ := json.MarshalIndent(r.Output, "", "  ")
			if r.Output == nil {
				combinedErr = multierror.Append(combinedErr, fmt.Errorf("got nil Output\nWant Output:%v", string(expectedOutputBytes)))
			} else {
				prettyPrint, err := PrettyPrintBeforeAfter(r.Output, *expectedOutput)
				if err != nil {
					combinedErr = multierror.Append(combinedErr, fmt.Errorf("error matching expected Output:%v", err))
				}
				if prettyPrint != "" {
					combinedErr = multierror.Append(combinedErr, fmt.Errorf("expected Output mismatch (-got, +want):%s", prettyPrint))
				}
			}
		} else {
			if r.Output != nil {
				outputBytes, _ := json.MarshalIndent(r.Output, "", "  ")
				combinedErr = multierror.Append(combinedErr, fmt.Errorf("got Output:%v\nWant nil Output", string(outputBytes)))
			}
		}
		return combinedErr
	})
	return TestCloudFrontFunctionWithCustomValidation(t, name, stage, event, validateResponse)
}

// TestCloudFrontFunctionWithCustomValidation performs a Function test and validate the response. Fails the test if there is an error.
func TestCloudFrontFunctionWithCustomValidation(t testing.TestingT, name string, stage string, event CloudFrontFunctionEvent, validateResponse responseValidator) error {
	functionStage := assertFunctionStage(t, stage)
	err := TestCloudFrontFunctionWithCustomValidationE(t, name, functionStage, event, validateResponse)
	require.NoError(t, err)
	return nil
}

// TestCloudFrontFunctionWithCustomValidationE performs a Function test and validate the response.
func TestCloudFrontFunctionWithCustomValidationE(t testing.TestingT, name string, stage types.FunctionStage, event CloudFrontFunctionEvent, validateResponse responseValidator) error {
	response, err := TestCloudFrontFunctionE(t, name, stage, event)
	if err != nil {
		return err
	}
	// log utilization for information purposes
	logger.Log(t, fmt.Sprintf("CloudFront Function Utilization: (%d%%)", response.Utilization))
	if err := validateResponse(response); err != nil {
		return CloudFrontFunctionValidationFailed{
			FunctionName: name + ":" + string(stage),
			Failures:     err,
		}
	}
	return nil
}

// TestCloudFrontFunctionE performs a Function test and validates the response.
func TestCloudFrontFunctionE(t testing.TestingT, name string, stage types.FunctionStage, event CloudFrontFunctionEvent) (*CloudFrontTestFunctionResult, error) {
	ctx := context.TODO()

	jsonData, err := json.Marshal(event)
	if err != nil {
		return nil, fmt.Errorf("error serializing CloudFront Function Event: %q", err)
	}

	client := NewCloudFrontclient(t)
	functionDetails, err := client.DescribeFunction(ctx, &cloudfront.DescribeFunctionInput{
		Name:  aws.String(name),
		Stage: stage,
	})
	if err != nil {
		return nil, err
	}

	r, err := client.TestFunction(ctx, &cloudfront.TestFunctionInput{
		EventObject: jsonData,
		IfMatch:     functionDetails.ETag,
		Name:        aws.String(name),
		Stage:       stage,
	})
	if err != nil {
		return nil, err
	}

	return parseTestResult(r.TestResult)
}

// NewCloudFrontclient returns a client for CloudFront. This will fail the test and
// stop execution if there is an error.
func NewCloudFrontclient(t testing.TestingT) *cloudfront.Client {
	sess, err := NewCloudFrontclientE(t)
	require.NoError(t, err)
	return sess
}

// NewCloudFrontclientE returns a client for CloudFront.
func NewCloudFrontclientE(t testing.TestingT) (*cloudfront.Client, error) {
	cfg, err := config.LoadDefaultConfig(context.TODO(), config.WithRegion("us-east-1"))
	if err != nil {
		return nil, err
	}
	return cloudfront.NewFromConfig(cfg), nil
}

// assertFunctionStage validates the function stage or fails the test.
func assertFunctionStage(t testing.TestingT, stage string) types.FunctionStage {
	var functionStage types.FunctionStage
	switch stage {
	case string(types.FunctionStageDevelopment):
		functionStage = types.FunctionStageDevelopment
	case string(types.FunctionStageLive):
		functionStage = types.FunctionStageLive
	default:
		t.Errorf("Invalid function stage: %s", stage)
		return types.FunctionStageDevelopment
	}
	return functionStage
}

func parseTestResult(testResult *types.TestResult) (*CloudFrontTestFunctionResult, error) {
	result := CloudFrontTestFunctionResult{
		ExecutionLogs: testResult.FunctionExecutionLogs,
	}
	var err error

	if aws.ToString(testResult.FunctionErrorMessage) != "" {
		result.ErrorMessage = aws.String(invalidFunctionErrorPrefix + *testResult.FunctionErrorMessage)
	}

	if testResult.ComputeUtilization != nil {
		result.Utilization, err = strconv.Atoi(aws.ToString(testResult.ComputeUtilization))
		if err != nil {
			return &result, err
		}
	}
	var output interface{}
	if testResult.FunctionOutput != nil {
		err := json.Unmarshal([]byte(*testResult.FunctionOutput), &output)
		if err != nil {
			return &result, fmt.Errorf("JSON decode error: %w", err)
		}
	}
	result.Output = make(map[string]interface{})
	if output != nil {
		outputMap, ok := output.(map[string]interface{})
		if ok {
			if len(outputMap) != 0 {
				result.Output = outputMap
			}
		}
	}
	return &result, nil
}
