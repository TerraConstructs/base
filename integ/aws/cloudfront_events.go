package aws

import (
	"encoding/json"
	"fmt"
	"os"
)

type Version string

const Version1_0 Version = "1.0"

func (v *Version) UnmarshalJSON(data []byte) error {
	var versionStr string
	if err := json.Unmarshal(data, &versionStr); err != nil {
		return err
	}
	if versionStr != string(Version1_0) {
		return fmt.Errorf("unsupported version: %s", versionStr)
	}
	*v = Version(versionStr)
	return nil
}

func (v Version) MarshalJSON() ([]byte, error) {
	if v != Version1_0 {
		return nil, fmt.Errorf("unsupported version: %s", v)
	}
	return json.Marshal(string(v))
}

// CloudFrontFunctionEvent represents the event object that you provide to test a CloudFront function.
// https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/functions-event-structure.html
type CloudFrontFunctionEvent struct {
	Version  Version   `json:"version"` // The version field contains a string that specifies the version of the CloudFront Functions event object. The current version is 1.0.
	Context  Context   `json:"context"` // The context object contains contextual information about the event
	Viewer   Viewer    `json:"viewer"`
	Request  *Request  `json:"request,omitempty"`
	Response *Response `json:"response,omitempty"`
}

type Context struct {
	DistributionDomainName string `json:"distributionDomainName"` // The CloudFront domain name (for example, d111111abcdef8.cloudfront.net) of the distribution that's associated with the event.
	DistributionID         string `json:"distributionId"`
	EventType              string `json:"eventType"`
	RequestID              string `json:"requestId"`
}

type Viewer struct {
	IP string `json:"ip"`
}

type ValueObject map[string]ValueEntry

type ValueEntry struct {
	Value      string       `json:"value"`
	MultiValue []MultiValue `json:"multiValue,omitempty"`
}

type MultiValue struct {
	Value string `json:"value"`
}

type Request struct {
	Method      string      `json:"method"`
	URI         string      `json:"uri"`
	Querystring ValueObject `json:"querystring"`
	Headers     ValueObject `json:"headers"`
	Cookies     ValueObject `json:"cookies,omitempty"`
}

type Response struct {
	StatusCode        int             `json:"statusCode"`
	StatusDescription *string         `json:"statusDescription,omitempty"`
	Headers           *ValueObject    `json:"headers,omitempty"`
	Cookies           *ResponseCookie `json:"cookies,omitempty"`
}

type ResponseCookie map[string]ResponseCookieEntry

type ResponseCookieEntry struct {
	Value      string            `json:"value"`
	Attributes string            `json:"attributes"`
	MultiValue []MultiValueAttrs `json:"multiValue,omitempty"`
}

type MultiValueAttrs struct {
	Value      string `json:"value"`
	Attributes string `json:"attributes"`
}

func ReadCloudFrontEvent(path string) (*CloudFrontFunctionEvent, error) {
	f, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	var t = CloudFrontFunctionEvent{
		Version: "1.0",
		Context: Context{
			EventType: "viewer-request",
		},
		Viewer: Viewer{
			IP: "1.2.3.4",
		},
	}

	err = json.Unmarshal(f, &t)
	return &t, err
}
