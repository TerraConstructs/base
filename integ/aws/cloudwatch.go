package aws

import (
	"fmt"
	"time"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/service/cloudwatchevents"
	"github.com/aws/aws-sdk-go/service/cloudwatchlogs"
	terratestaws "github.com/gruntwork-io/terratest/modules/aws"
	"github.com/gruntwork-io/terratest/modules/retry"
	"github.com/gruntwork-io/terratest/modules/testing"
	"github.com/stretchr/testify/require"
)

// WaitForLogEvents waits for log events to appear in the given CloudWatch Log group in the given region
func WaitForLogEvents(
	t testing.TestingT,
	awsRegion string,
	logGroupName string,
	maxRetries int,
	sleepBetweenRetries time.Duration,
) []string {
	events, err := WaitForLogEventsE(t, awsRegion, logGroupName, maxRetries, sleepBetweenRetries)
	require.NoError(t, err)
	return events
}

// WaitForLogEventsE waits for log events to appear in the given CloudWatch Log group in the given region
func WaitForLogEventsE(
	t testing.TestingT,
	awsRegion string,
	logGroupName string,
	maxRetries int,
	sleepBetweenRetries time.Duration,
) ([]string, error) {
	var result []string

	description := fmt.Sprintf("Waiting for log events in log group %s", logGroupName)

	_, err := retry.DoWithRetryE(
		t,
		description,
		maxRetries,
		sleepBetweenRetries,
		func() (string, error) {
			messages, err := FilterLogEventsE(t, awsRegion, logGroupName)
			if err != nil {
				return "", err
			}

			if len(messages) > 0 {
				result = messages
				return "Log events found", nil
			} else {
				return "", fmt.Errorf("no log events found yet")
			}
		},
	)
	if err != nil {
		return nil, err
	}
	return result, nil
}

// GetCloudWatchLogEntries returns the CloudWatch log messages in the given region for the given log stream and log group.
func FilterLogEvents(t testing.TestingT, awsRegion string, logGroupName string) []string {
	out, err := FilterLogEventsE(t, awsRegion, logGroupName)
	if err != nil {
		t.Fatal(err)
	}
	return out
}

// GetCloudWatchLogEntriesE returns the CloudWatch log messages in the given region for the given log stream and log group.
func FilterLogEventsE(t testing.TestingT, awsRegion string, logGroupName string) ([]string, error) {
	client, err := terratestaws.NewCloudWatchLogsClientE(t, awsRegion)
	if err != nil {
		return nil, err
	}

	output, err := client.FilterLogEvents(&cloudwatchlogs.FilterLogEventsInput{
		LogGroupName: aws.String(logGroupName),
	})

	if err != nil {
		return nil, err
	}

	entries := []string{}
	for _, event := range output.Events {
		entries = append(entries, *event.Message)
	}

	return entries, nil
}

// DescribeEventRule returns the details of the specified rule.
func DescribeEventRule(t testing.TestingT, awsRegion string, ruleName string) *CloudwatchEventsRuleInfo {
	out, err := DescribeEventRuleE(t, awsRegion, ruleName)
	if err != nil {
		t.Fatal(err)
	}
	return out
}

type CloudwatchEventsRuleInfo struct {
	Name               string // The name of the rule.
	Description        string // The description of the rule.
	State              string // Specifies whether the rule is enabled or disabled.
	EventPattern       string // The event pattern.
	ScheduleExpression string // The scheduling expression. For example, "cron(0 20 * * ? *)", "rate(5 minutes)".
}

// DescribeEventRuleE returns the details of the specified rule.
func DescribeEventRuleE(t testing.TestingT, awsRegion string, ruleName string) (*CloudwatchEventsRuleInfo, error) {
	client, err := NewCloudWatchEventsClientE(t, awsRegion)
	if err != nil {
		return nil, err
	}

	output, err := client.DescribeRule(&cloudwatchevents.DescribeRuleInput{
		Name: aws.String(ruleName),
	})

	if err != nil {
		return nil, err
	}
	ruleInfo := &CloudwatchEventsRuleInfo{
		Name:               aws.StringValue(output.Name),
		Description:        aws.StringValue(output.Description),
		State:              aws.StringValue(output.State),
		EventPattern:       aws.StringValue(output.EventPattern),
		ScheduleExpression: aws.StringValue(output.ScheduleExpression),
	}
	return ruleInfo, nil
}

// NewCloudWatchEventsClient creates a new CloudWatch Events client.
func NewCloudWatchEventsClient(t testing.TestingT, region string) *cloudwatchevents.CloudWatchEvents {
	client, err := NewCloudWatchEventsClientE(t, region)
	if err != nil {
		t.Fatal(err)
	}
	return client
}

// NewCloudWatchEventsClientE creates a new CloudWatch Logs client.
func NewCloudWatchEventsClientE(t testing.TestingT, region string) (*cloudwatchevents.CloudWatchEvents, error) {
	sess, err := terratestaws.NewAuthenticatedSession(region)
	if err != nil {
		return nil, err
	}
	return cloudwatchevents.New(sess), nil
}
