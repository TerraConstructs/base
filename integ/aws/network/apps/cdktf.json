{
  "//": "Dummy file to satisfy the bundling findUp logic"
}
