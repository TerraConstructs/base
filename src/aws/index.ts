// core
export * from "./arn";
export * from "./beacon";
export * from "./spec";

// core JSII compatible generated interfaces
export * from "./provider-config.generated";

// classes
// export * as iam from "./iam"; // Iam floyd is not JSII compatible
export * as compute from "./compute";
export * as network from "./network";
export * as staticsite from "./staticsite";
