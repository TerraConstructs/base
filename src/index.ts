export * from "./construct-base";
export * from "./stack-base";
export * from "./duration";
export * from "./size";

// cloud specific exports
export * as aws from "./aws";
