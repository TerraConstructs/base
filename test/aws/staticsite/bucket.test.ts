import path from "path";
import { Testing } from "cdktf";
import "cdktf/lib/testing/adapters/jest";
import { staticsite, AwsSpec } from "../../../src/aws";

const environmentName = "Test";
const gridUUID = "123e4567-e89b-12d3";
const providerConfig = { region: "us-east-1" };
describe("Bucket", () => {
  test("Should synth and match SnapShot", () => {
    // GIVEN
    const spec = getAwsSpec();
    // WHEN
    new staticsite.Bucket(spec, "HelloWorld", {
      namePrefix: "hello-world",
      path: path.join(__dirname, "fixtures", "site"),
      websiteConfig: {
        enabled: true,
      },
      public: true,
    });
    // THEN
    expect(Testing.synth(spec)).toMatchSnapshot();
  });
});

function getAwsSpec(): AwsSpec {
  const app = Testing.app();
  return new AwsSpec(app, "TestSpec", {
    environmentName,
    gridUUID,
    providerConfig,
  });
}
