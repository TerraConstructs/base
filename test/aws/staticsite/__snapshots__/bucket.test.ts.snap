// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bucket Should synth and match SnapShot 1`] = `
"{
  "data": {
    "aws_caller_identity": {
      "CallerIdentity": {
        "provider": "aws"
      }
    },
    "aws_partition": {
      "Partitition": {
        "provider": "aws"
      }
    }
  },
  "provider": {
    "aws": [
      {
        "region": "us-east-1"
      }
    ]
  },
  "resource": {
    "aws_s3_bucket": {
      "HelloWorld_7964D1E8": {
        "bucket_prefix": "123e4567-e89b-12d3-hello-world",
        "tags": {
          "Name": "Test-HelloWorld",
          "et:grid:EnvironmentName": "Test",
          "et:grid:UUID": "123e4567-e89b-12d3"
        }
      }
    },
    "aws_s3_bucket_acl": {
      "HelloWorld_PublicAcl_A32D5A72": {
        "acl": "public-read",
        "bucket": "\${aws_s3_bucket.HelloWorld_7964D1E8.bucket}",
        "depends_on": [
          "aws_s3_bucket_ownership_controls.HelloWorld_OwnershipControls_179F29CE",
          "aws_s3_bucket_public_access_block.HelloWorld_PublicAccessBlock_8B1865DA"
        ]
      }
    },
    "aws_s3_bucket_ownership_controls": {
      "HelloWorld_OwnershipControls_179F29CE": {
        "bucket": "\${aws_s3_bucket.HelloWorld_7964D1E8.bucket}",
        "rule": {
          "object_ownership": "BucketOwnerPreferred"
        }
      }
    },
    "aws_s3_bucket_policy": {
      "HelloWorld_Policy_351F392A": {
        "bucket": "\${aws_s3_bucket.HelloWorld_7964D1E8.bucket}",
        "depends_on": [
          "aws_s3_bucket_public_access_block.HelloWorld_PublicAccessBlock_8B1865DA"
        ],
        "policy": "{\\"Version\\":\\"2012-10-17\\",\\"Statement\\":[{\\"Action\\":\\"s3:GetObject\\",\\"Resource\\":\\"\${aws_s3_bucket.HelloWorld_7964D1E8.arn}/*\\",\\"Effect\\":\\"Allow\\",\\"Principal\\":{\\"AWS\\":[\\"*\\"]}}]}"
      }
    },
    "aws_s3_bucket_public_access_block": {
      "HelloWorld_PublicAccessBlock_8B1865DA": {
        "block_public_acls": false,
        "block_public_policy": false,
        "bucket": "\${aws_s3_bucket.HelloWorld_7964D1E8.bucket}",
        "ignore_public_acls": false,
        "restrict_public_buckets": false
      }
    },
    "aws_s3_bucket_website_configuration": {
      "HelloWorld_WebsiteConfig_72ADC681": {
        "bucket": "\${aws_s3_bucket.HelloWorld_7964D1E8.bucket}",
        "index_document": {
          "suffix": "index.html"
        }
      }
    },
    "aws_s3_object": {
      "HelloWorld_File--images--officespace-greatjpg_34A1DCCE": {
        "bucket": "\${aws_s3_bucket.HelloWorld_7964D1E8.bucket}",
        "content_type": "image/jpeg",
        "key": "/images/officespace-great.jpg",
        "source": "assets/HelloWorld_PathAsset_4757FC2F/7C1CC84268AB29AAED03995506EAD038/images/officespace-great.jpg",
        "source_hash": "\${filemd5(\\"assets/HelloWorld_PathAsset_4757FC2F/7C1CC84268AB29AAED03995506EAD038/images/officespace-great.jpg\\")}",
        "tags": {
          "Name": "Test-HelloWorld",
          "et:grid:EnvironmentName": "Test",
          "et:grid:UUID": "123e4567-e89b-12d3"
        }
      },
      "HelloWorld_File--indexhtml_5799AAE8": {
        "bucket": "\${aws_s3_bucket.HelloWorld_7964D1E8.bucket}",
        "content_type": "text/html; charset=utf-8",
        "key": "/index.html",
        "source": "assets/HelloWorld_PathAsset_4757FC2F/7C1CC84268AB29AAED03995506EAD038/index.html",
        "source_hash": "\${filemd5(\\"assets/HelloWorld_PathAsset_4757FC2F/7C1CC84268AB29AAED03995506EAD038/index.html\\")}",
        "tags": {
          "Name": "Test-HelloWorld",
          "et:grid:EnvironmentName": "Test",
          "et:grid:UUID": "123e4567-e89b-12d3"
        }
      }
    }
  },
  "terraform": {
    "required_providers": {
      "aws": {
        "source": "aws",
        "version": "5.60.0"
      }
    }
  }
}"
`;
